lib/Common/Memory/Recycler.cpp

```diff
@@ -1098,20 +1098,31 @@ Recycler::Prime()
     });
 }
 
-void
-Recycler::AddExternalMemoryUsage(size_t size)
+#define KB64  (64 * 1024)
+#define KB128 (2  * KB64)
+bool Recycler::RequestExternalMemoryAllocation(size_t size, bool collectOnFail)
 {
-    this->autoHeap.uncollectedAllocBytes += size;
-    this->autoHeap.uncollectedExternalBytes += size;
-    // Generally normal GC can cleanup the uncollectedAllocBytes. But if external components
-    // do fast large allocations in a row, normal GC might not kick in. Let's force the GC
-    // here if we need to collect anyhow.
-    CollectNow<CollectOnAllocation>();
-}
+    bool success = recyclerPageAllocator.RequestAlloc(size);
+    // if this is not a success, caller will be collecting aggresively anyways
+    // otherwise caller should set collectOnFail
+    if (success || collectOnFail)
+    {
+        if (success)
+        {
+            this->autoHeap.uncollectedExternalBytes += size;
+        }
 
-bool Recycler::RequestExternalMemoryAllocation(size_t size)
-{
-    return recyclerPageAllocator.RequestAlloc(size);
+        // Generally normal GC can cleanup the uncollectedAllocBytes. But if external components
+        // do fast large allocations in a row, normal GC might not kick in. Let's force the GC
+        // here if we need to collect anyhow.
+
+        if (collectOnFail || this->autoHeap.uncollectedExternalBytes > KB128 || size >= KB64)
```

Comment:
so, if I'm reading this correctly, if ever we end up in a place where `uncollectedExternalBytes is > 128KB`, every call to this function will trigger a GC.  Is there some way we can avoid this?  E.g., is there an API to "suggest a GC", and let some centralized GC logic worry about when to schedule that?

Reply:
No it won’t trigger GC. It will check whether it’s the time to do GC by checking other heuristics as is. If it really collects, well, uncollectedExternalBytes will be set to 0

Reply:
sorry, not following why `uncollectedExternalBytes` will be 0.  Won't they only be 0 if GC could actually collect those?

Reply:
How do we plan to do that? (Since some of it could be really external) and the only reason for this parameter to be exist is that we could measure the external memory pressure and try GC

Reply:
sorry, not following.  I have two questions:

1.  What is `uncollectedExternalBytes` and when does it go to 0?
2.  Are we ever subject to a condition where `uncollectedExternalBytes > KB128` and calling `CollectNow` will trigger GCs that are unable to reduce `uncollectedExternalBytes`?  (i.e., does this change subject us to overly-aggressive GC?)

Reply:
`uncollectedExternalBytes` symbolize an extra pressure number. i.e. host app allocated a piece of memory and notified us with the size.

At the moment, we do try to GC, regardless what `uncollectedExternalBytes` is..

So what I did is not limited to putting a `128KB` threshold. Whenever `requestAlloc` fails, we do actually try to GC, so adding up `uncollectedExternalBytes` and trying again was waste of cpu tick. Cleaned up there a bit and put a threshold.